pkg/reconciler/namespacecontroller/namespace_predicate.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"k8s.io/klog/v2"
+	"kpt.dev/configsync/pkg/api/configmanagement"
+	"kpt.dev/configsync/pkg/syncer/differ"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+const (
+	metadataNameLabel = "kubernetes.io/metadata.name"
+)
+
+var _ predicate.Predicate = NamespaceLabelChangedPredicate{}
+
+// NamespaceLabelChangedPredicate triggers a new reconciliation when Namespace
+// labels are changed, so NamespaceSelector can select a different set of resources.
+type NamespaceLabelChangedPredicate struct {
+	predicate.Funcs
+}
+
+// canTriggerReconcile returns whether the Namespace has labels that can trigger
+// a reconciliation.
+// It can't trigger a reconciliation for the following scenarios:
+//   - If it only contains the `kubernetes.io/metadata.name` label. The Kubernetes
+//     control plane sets the immutable label `kubernetes.io/metadata.name` on all
+//     namespaces, the value is the same as Namespace name. If a Namespace only
+//     has that label, there is no need to use NamespaceSelector. It reduces the
+//     number of reconciliations by eliminating this scenario.
+//   - If it is a CS controller Namespace, it should not trigger a reconciliation
```

Comment:
Didn't we recently have a customer issue where they had added config-management-system to their source of truth and it was managing the labels or something?

Is this a hard-coded constraint somewhere?

Reply:
We never allow managing `config-management-system`: https://cloud.google.com/anthos-config-management/docs/reference/errors#knv1034. 

https://github.com/GoogleContainerTools/kpt-config-sync/pull/961 recently extends the constraints to config-management-monitoring and resource-group-controller namespaces. IIRC, the customer issue was adding config-management-monitoring to the SoT.